test/python/test_jacobian.py

```diff
@@ -702,3 +870,275 @@ def setUpClass(cls):
+    equation = "H2O(S) => H2O + PT(S)"
+    rate_type = "interface-Arrhenius"
+
+class SurfaceFullTests:
+    # Generic test class to check derivatives evaluated for an entire reaction mechanisms
+    rtol = 1e-4
+
+    @classmethod
+    def setUpClass(cls):
+        # all species indexs
+        all_species = cls.surf.species() + cls.gas.species()
+        cls.sidxs  = {spec.name:i for i, spec in enumerate(all_species)}
+
+    def setUp(self):
+        # gas phase
+        self.gas.TPX = self.gas_tpx
+        self.gas.derivative_settings = {} # reset defaults
+        # surface phase
+        self.surf.TPX = self.surf_tpx
+        self.surf.derivative_settings = {"skip-coverage-dependence": True, "skip-electrochemistry": True}
+
+    # closure to get concentrations vector
+    def get_concentrations(self):
+        return np.concatenate((self.surf.concentrations, self.gas.concentrations))
+
+    def test_forward_rop_ddCi(self):
+        # check forward rop against analytical result
+        drop = self.surf.forward_rates_of_progress_ddCi
+        rop = self.surf.forward_rates_of_progress
+        conc = self.get_concentrations()
+        # multiply derivatives with concentrations
+        drop = np.matmul(drop, conc)
+        # get total reactant reaction orders
+        total_orders = []
+        for rxn in self.surf.reactions():
+            orders = rxn.orders
+            curr_order = 0
+            for k, v in rxn.reactants.items():
+                if k in orders:
+                    curr_order += orders[k]
+                else:
+                    curr_order += v
+            total_orders.append(curr_order)
+        total_orders = np.array(total_orders)
+        # divide derivatives by total orders which should provide rops
+        drop /= total_orders
+        # compare the two
+        self.assertArrayNear(drop, rop, self.rtol)
+
+    def test_reverse_rop_ddCi(self):
+        # check forward rop against analytical result
+        drop = self.surf.reverse_rates_of_progress_ddCi
+        rop = self.surf.reverse_rates_of_progress
+        conc = self.get_concentrations()
+        # multiply derivatives with concentrations
+        drop = np.matmul(drop, conc)
+        # get total reactant reaction orders
+        total_orders = []
+        for rxn in self.surf.reactions():
+            orders = rxn.orders
+            curr_order = 0
+            for k, v in rxn.products.items():
+                if k in orders:
+                    curr_order += orders[k]
+                else:
+                    curr_order += v
+            total_orders.append(curr_order)
+        total_orders = np.array(total_orders)
+        # divide derivatives by total orders which should provide rops
+        drop /= total_orders
+        # compare the two
+        self.assertArrayNear(drop, rop, self.rtol)
+
+    def test_net_rop_ddCi(self):
+        # check forward rop against analytical result
+        drop = self.surf.net_rates_of_progress_ddCi
```

Comment:
How does this test succeed without any reference to `rxn.products`? I don't think the logic by which the forward and reverse rate deriviatives are calculated extends to the net rates. The comment here seems incorrect, also.